### PR TITLE
My Account > Change account information and Newsletter subscription

### PR DESCRIPTION
**Pull request by nuzil**
### Description

Implementing 2 issues: 

Customer account edit data and customer newsletter subscription

https://app.zenhub.com/workspace/o/magento/graphql-ce/issues/56
https://app.zenhub.com/workspace/o/magento/graphql-ce/issues/55

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/graphql-ce?pullRequest=162) <br/>All committers have signed the CLA.

File: app/code/Magento/CustomerGraphQl/Model/Resolver/Customer/CustomerDataProvider.php
Comment (naydav) [inline app/code/Magento/CustomerGraphQl/Model/Resolver/Customer/CustomerDataProvider.php]: Pls, use `private` access

File: app/code/Magento/CustomerGraphQl/Model/Resolver/Customer/CustomerDataProvider.php
Comment (naydav) [inline app/code/Magento/CustomerGraphQl/Model/Resolver/Customer/CustomerDataProvider.php]: Why do we need mathod-wrapper over Repository?

File: app/code/Magento/CustomerGraphQl/Model/Resolver/Customer/CustomerDataProvider.php
Comment (naydav) [inline app/code/Magento/CustomerGraphQl/Model/Resolver/Customer/CustomerDataProvider.php:14]: Looks like this class has a lot of responsibility (looks like helper in Magento 1)
Need to split the class

File: app/code/Magento/CustomerGraphQl/Model/Resolver/Customer/CustomerDataProvider.php
Comment (naydav) [inline app/code/Magento/CustomerGraphQl/Model/Resolver/Customer/CustomerDataProvider.php]: You can simplify this condition

File: app/code/Magento/CustomerGraphQl/Model/Resolver/CustomerUpdate.php
Comment (naydav) [inline app/code/Magento/CustomerGraphQl/Model/Resolver/CustomerUpdate.php]: Do we need to return something if it is 'command' operation?

File: app/code/Magento/CustomerGraphQl/etc/schema.graphqls
Comment (naydav) [inline app/code/Magento/CustomerGraphQl/etc/schema.graphqls]: Also, need to cover functionality with API-functional tests
Now we have a lot of examples and tests writing is very simple

**Review by naydav — CHANGES_REQUESTED** [review]

File: app/code/Magento/CustomerGraphQl/Model/Resolver/Customer/CustomerDataProvider.php
Comment (nuzil) [reply app/code/Magento/CustomerGraphQl/Model/Resolver/Customer/CustomerDataProvider.php]: Good catch, I planned to use it in another location, but currently it not needed anymore

File: app/code/Magento/CustomerGraphQl/Model/Customer/ChangeSubscriptionStatus.php
Comment (paliarush) [inline app/code/Magento/CustomerGraphQl/Model/Customer/ChangeSubscriptionStatus.php:40]: Does not make sense to load records from DB if the next two conditions fail.

File: app/code/Magento/CustomerGraphQl/Model/Customer/CheckCustomerAccount.php
Comment (paliarush) [inline app/code/Magento/CustomerGraphQl/Model/Customer/CheckCustomerAccount.php:67]: Usually it is better to use 'is not' instead of 'isn't' in error message.

File: app/code/Magento/CustomerGraphQl/Model/Customer/CheckCustomerAccount.php
Comment (paliarush) [inline app/code/Magento/CustomerGraphQl/Model/Customer/CheckCustomerAccount.php:64]: Why customer type is needed here? For guests $customerID will be 0.
Also, why these arguments are optional? We can just ask for (int)$customerId

File: app/code/Magento/CustomerGraphQl/Model/Customer/CheckCustomerPassword.php
Comment (paliarush) [inline app/code/Magento/CustomerGraphQl/Model/Customer/CheckCustomerPassword.php:17]: We should use nouns as class names. Please consider changing it to CustomerPasswordVerifier or CustomerAuthenticator.

Please fix the names of all affected variables which hold this object as well.

File: app/code/Magento/CustomerGraphQl/Model/Customer/UpdateAccountInformation.php
Comment (keharper) [inline app/code/Magento/CustomerGraphQl/Model/Customer/UpdateAccountInformation.php:76]: Change exception message to  'Provide the current "password" to change "email".'

File: app/code/Magento/CustomerGraphQl/Model/Resolver/ChangePassword.php
Comment (keharper) [inline app/code/Magento/CustomerGraphQl/Model/Resolver/ChangePassword.php:73]: Change exception message to 'Specify the "currentPassword" value.`

File: app/code/Magento/CustomerGraphQl/Model/Resolver/ChangePassword.php
Comment (keharper) [inline app/code/Magento/CustomerGraphQl/Model/Resolver/ChangePassword.php:77]: Change exception message to 'Specify the "newPassword" value.'

File: app/code/Magento/CustomerGraphQl/Model/Resolver/GenerateCustomerToken.php
Comment (keharper) [inline app/code/Magento/CustomerGraphQl/Model/Resolver/GenerateCustomerToken.php:48]: Change exception message to 'Specify the "email"  value.'

File: app/code/Magento/CustomerGraphQl/Model/Resolver/GenerateCustomerToken.php
Comment (keharper) [inline app/code/Magento/CustomerGraphQl/Model/Resolver/GenerateCustomerToken.php:52]: Change exception message to 'Specify the "password"  value.'

File: app/code/Magento/CustomerGraphQl/Model/Resolver/UpdateCustomer.php
Comment (keharper) [inline app/code/Magento/CustomerGraphQl/Model/Resolver/UpdateCustomer.php:73]: Change exception message to 'Specify the "input"  value.'

File: app/code/Magento/CustomerGraphQl/etc/schema.graphqls
Comment (keharper) [inline app/code/Magento/CustomerGraphQl/etc/schema.graphqls:9]: Change the description text to "Retrieve the customer token"

File: app/code/Magento/CustomerGraphQl/etc/schema.graphqls
Comment (keharper) [inline app/code/Magento/CustomerGraphQl/etc/schema.graphqls:10]: Change the description text to "Changes the password for the logged-in customer"

File: app/code/Magento/CustomerGraphQl/etc/schema.graphqls
Comment (keharper) [inline app/code/Magento/CustomerGraphQl/etc/schema.graphqls:11]: Change the description text to "Update the customer's personal information"

File: app/code/Magento/CustomerGraphQl/etc/schema.graphqls
Comment (keharper) [inline app/code/Magento/CustomerGraphQl/etc/schema.graphqls:12]: Change the description text to the "Revoke the customer token"

**Review by keharper — CHANGES_REQUESTED** [review]

File: app/code/Magento/CustomerGraphQl/Model/Resolver/ChangePassword.php
Comment (paliarush) [inline app/code/Magento/CustomerGraphQl/Model/Resolver/ChangePassword.php:20]: We need explicit description for the purpose of this class.

File: app/code/Magento/CustomerGraphQl/Model/Resolver/ChangePassword.php
Comment (paliarush) [inline app/code/Magento/CustomerGraphQl/Model/Resolver/ChangePassword.php:27]: Same issue as with `checkCustomerPassword`, should be a noun.

File: app/code/Magento/CustomerGraphQl/Model/Resolver/UpdateCustomer.php
Comment (paliarush) [inline app/code/Magento/CustomerGraphQl/Model/Resolver/UpdateCustomer.php:72]: I believe this will be validated by the GraphQL schema. Please verify this and remove the check if it is already done by the framework. 

If this check can be eliminated, please check similar checks in other resolvers.

File: app/code/Magento/CustomerGraphQl/composer.json
Comment (paliarush) [inline app/code/Magento/CustomerGraphQl/composer.json:9]: Please remember to fix health index etalon tests

File: app/code/Magento/CustomerGraphQl/etc/schema.graphqls
Comment (paliarush) [inline app/code/Magento/CustomerGraphQl/etc/schema.graphqls:11]: Must be marked as required `UpdateCustomerInput!`

File: app/code/Magento/CustomerGraphQl/etc/schema.graphqls
Comment (paliarush) [inline app/code/Magento/CustomerGraphQl/etc/schema.graphqls:12]: Let's create a wrapper for the revoke result to allow for future extensibility.

File: lib/internal/Magento/Framework/GraphQl/Exception/GraphQlAlreadyExistsException.php
Comment (paliarush) [inline lib/internal/Magento/Framework/GraphQl/Exception/GraphQlAlreadyExistsException.php:15]: More meaningful description is required for this and other newly introduced exceptions.

File: lib/internal/Magento/Framework/GraphQl/Query/ResolverInterface.php
Comment (paliarush) [inline lib/internal/Magento/Framework/GraphQl/Query/ResolverInterface.php:23]: Is fixed in 2.3.0 performance improvements and will result in a conflict. Better revert it here.

**Review by paliarush — CHANGES_REQUESTED** [review]

**Comment by naydav** [comment]
All CR fixes will be provided in separate PR

**Pull request merged**
